be/src/exec/pipeline/scan/morsel.h

```diff
@@ -98,5 +120,84 @@ class FixedMorselQueue final : public MorselQueue {
     std::atomic<size_t> _pop_index;
 };
 
+class PhysicalSplitMorselQueue final : public MorselQueue {
+public:
+    explicit PhysicalSplitMorselQueue(Morsels&& morsels, int64_t degree_of_parallelism, int64_t splitted_scan_rows)
+            : _morsels(std::move(morsels)),
+              _num_original_morsels(_morsels.size()),
+              _degree_of_parallelism(degree_of_parallelism),
+              _splitted_scan_rows(splitted_scan_rows) {}
+    ~PhysicalSplitMorselQueue() override = default;
+
+    std::vector<TInternalScanRange*> olap_scan_ranges() const override;
+
+    void set_key_ranges(const std::vector<OlapScanRange*>& key_ranges) override;
+    void set_tablets(const std::vector<TabletSharedPtr>& tablets) override { _tablets = tablets; }
+    void set_tablet_rowsets(const std::vector<std::vector<RowsetSharedPtr>>& tablet_rowsets) override {
+        _tablet_rowsets = tablet_rowsets;
+    }
+
+    size_t num_morsels() const override { return _degree_of_parallelism; }
+    bool empty() const override { return _tablet_idx >= _tablets.size(); }
+    StatusOr<MorselPtr> try_get() override;
+
+    std::string name() const override { return "physical_split_morsel_queue"; }
+
+    // Whether DOP after splitting tablets is greater than the number of tablets.
+    // Only used by broadcast HashJoinProbeOperator. It will insert local exchange
+    // to balance the chunk from scan node and increase parallelism of the probe operators,
+    // if the number of tablets is less than DOP.
+    bool need_rebalance() const override { return _num_original_morsels < _degree_of_parallelism; }
+
+private:
+    static rowid_t _lower_bound_ordinal(Segment* segment, const vectorized::SeekTuple& key, bool lower);
+    static rowid_t _upper_bound_ordinal(Segment* segment, const vectorized::SeekTuple& key, bool lower, rowid_t end);
+
+    ScanMorsel* _cur_scan_morsel();
+    BetaRowset* _cur_rowset();
+    // Return nullptr, when _segment_idx exceeds the segments of the current rowset.
+    Segment* _cur_segment();
+
+    // Returning false means that there is no more segment to read.
+    bool _next_segment();
+    // Load the meta of the new rowset and the index of the new segment,
+    // and find the rowid range of each key range in this segment.
+    Status _init_segment();
+
+private:
+    std::mutex _mutex;
+
+    const Morsels _morsels;
+    // The number of the morsels before split them to pieces.
+    const size_t _num_original_morsels;
+    const int64_t _degree_of_parallelism;
+    // The minimum number of rows picked up from a segment at one time.
+    const int64_t _splitted_scan_rows;
+
+    /// Key ranges passed to the storage layer.
+    // Possible values are "gt", "ge", "eq".
+    std::string _range_start_op = "";
```

Comment:
why not use an enum?

Reply:
The storage layer use `std::string`, `_range_start_op` is used to pass it to storage layer.